utilities/transactions/write_unprepared_txn.h

```diff
@@ -199,6 +208,49 @@ class WriteUnpreparedTxn : public WritePreparedTxn {
   // but in some cases, we should be able to restore the previously largest
   // value when calling RollbackToSavepoint.
   SequenceNumber largest_validated_seq_;
+
+  struct SavePoint {
+    // Record of unprep_seqs_ at this savepoint. The set of unprep_seq is
+    // used during RollbackToSavepoint to determine visibility when restoring
+    // old values.
+    //
+    // Since all unprep_seqs_ sets further down the stack must be subsets, this
+    // can potentially deduplicated by just storing set difference.
+    std::map<SequenceNumber, size_t> unprep_seqs_;
+
+    SavePoint(const std::map<SequenceNumber, size_t>& seqs)
+        : unprep_seqs_(seqs){};
+  };
+
+  // We have 3 data structures holding savepoint information:
+  // 1. TransactionBaseImpl::save_points_
+  // 2. WriteUnpreparedTxn::wup_save_points_
+  // 3. WriteUnpreparecTxn::save_point_boundaries_
+  //
+  // TransactionBaseImpl::save_points_ holds information about all write
+  // batches, including the current in-memory write_batch_, or unprepared
+  // batches that have been written out. Its responsibility is just to track
+  // which keys have been modified in every savepoint.
+  //
+  // WriteUnpreparedTxn::wup_save_points_ holds information about savepoints set
+  // on unprepared batches that have already flushed. It just holds the
+  // unprep_seqs_ at that savepoint, so that the rollback process can determine
+  // which keys were visible at that point in time.
+  //
+  // WriteUnpreparecTxn::save_point_boundaries_ holds information about
+  // savepoints on the current in-memory write_batch_. It simply records the
+  // size of the write batch at every savepoint. This information is actually
```

Comment:
The line after "This information" could be replaced with a simple TODO:
// TODO(lth): remove the redundancy between save_point_boundaries_ and write_batch_.save_points_